### PR TITLE
chore(Scripts/Karazhan): rework Shade of Aran using TaskScheduler and fixes

### PR DESCRIPTION
**Pull request by elthehablo**
## Changes Proposed:
-  Use TaskScheduler instead of old update with diffs
-  Fix some bugs
- make water elementals unable to move for the time being

## Issues Addressed:

- Progress https://github.com/chromiecraft/chromiecraft/issues/4892
- Closes https://github.com/chromiecraft/chromiecraft/issues/5600 (but in a hacky way, but this is a general problem with caster AI on AC so this is perhaps the best solution here)

## Tests Performed:

- 
- 

## How to Test the Changes:

1. Go to Kara
2. Fight Aran
3.

Note: his mana doesn't really get depleted that quickly. Intentional? How does this work on retail?

Add the following to more easily test the drinking mechanic (to the regular scheduler):

```cpp
).Schedule(15s, [this](TaskContext)
{
	me->Yell("Oof ouch owie, my mana!", LANG_UNIVERSAL);
	me->SetPower(POWER_MANA, 3000);
});
```

## Known Issues and TODO List:

- [ ] movement is wonky. elementals should primarily stay still like casters, but on AC they move. this issue is hard to fix because it is tied to a more general problem with behaviour for casters
- [ ] timers with sniffs
The issues the first TODO creates are the following:
- https://github.com/chromiecraft/chromiecraft/issues/5600
- https://github.com/chromiecraft/chromiecraft/issues/4891

I want to fix this issue in this PR too:
https://github.com/chromiecraft/chromiecraft/issues/4892
With the latest commits, it works quite nicely. <del>I think he's not sitting properly as he's stil chasing his victim. Perhaps we can turn this off, but we do not want to turn off combat, so there might be some issue there. I can try something there</del>. Sitting works!

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Nyeriah** [comment]
Hey, thanks for the PR!

Here's a quick review:

- You can use an array for the positions and them iterate through the array to spawn the elementals (gotta fix this for merge)
- CREATURE_* should instead be NPC_* as that's the naming convention
- However, there's iirc a spell for the elementals (not required in this PR, we can do the research later, I have some work on this)
- The SetFaction() calls are probably all wrong as faction rarely ever changes during scripts
- Some other flags used to root things are also likely not correct

File: src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp [hunk 1]
```diff
@@ -70,560 +71,492 @@ enum SuperSpell
...
+                Unit* target = SelectTarget(SelectTargetMethod::Random, 1, 100, true);
+                if (!target)
+                    return;
+
+                DoStartNoMovement(target);
+                ElementalTwo->SetInCombatWithZone();
+                ElementalTwo->CombatStart(target);
+                ElementalTwo->SetFaction(me->GetFaction());
+                ElementalTwo->SetUnitMovementFlags(MOVEMENTFLAG_ROOT);
+                ElementalTwo->SetModifierValue(UNIT_MOD_RESISTANCE_FROST, BASE_VALUE, 0);
```
Comment (Nyeriah) [inline src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp:151]: Removing resistance? should be done in DB
Root flags also in DB

File: src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp [hunk 2]
```diff
@@ -70,560 +71,492 @@ enum SuperSpell
...
-        }
+        instance->SetData(DATA_ARAN, DONE);
 
-        void UpdateAI(uint32 diff) override
+        if (GameObject* libraryDoor = instance->instance->GetGameObject(instance->GetGuidData(DATA_GO_LIBRARY_DOOR)))
         {
-            if (!UpdateVictim())
-                return;
+            libraryDoor->SetGoState(GO_STATE_ACTIVE);
+            libraryDoor->RemoveGameObjectFlag(GO_FLAG_NOT_SELECTABLE);
```
Comment (Nyeriah) [inline src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp:224]: I think it changes its status to locked, Karazhan doors have weird behaviors

File: src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp (hunk 1 above)
Comment (Kitzunu) [reply src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp:151]: Base resistance is sniffed, no?

**Comment by Kitzunu** [comment]
All good points Nyeriah has!
I will just post this to give you pointers to the c++ standard https://www.azerothcore.org/wiki/cpp-code-standards#array-of-structs

Thanks for the PR!

File: src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp (hunk 1 above)
Comment (elthehablo) [reply src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp:151]: Do you want me to actually add the root flags in DB as well? As they may be incorrect, and didn't seem to work (well) regardless

**Comment by elthehablo** [comment]
> All good points Nyeriah has! I will just post this to give you pointers to the c++ standard https://www.azerothcore.org/wiki/cpp-code-standards#array-of-structs
> 
> Thanks for the PR!

Thanks matey. I assume Sunwell or wherever this script is originally from didn't have these standards lol

File: src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp (hunk 1 above)
Comment (elthehablo) [reply src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp:151]: Just checked. DB has Frost immunity for these. So should be sufficient to just remove this

File: src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp (hunk 2 above)
Comment (elthehablo) [reply src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp:220]: I agree. Also the issue with Malchezaar door

**Comment by elthehablo** [comment]
I need to fix 1 more thing and see if it's not completely broken with some tests

File: src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp [hunk 3]
```diff
@@ -77,6 +77,14 @@ enum Groups
     GROUP_DRINKING      = 1
 };
 
+float elementalPositions[4][4] =
```
Comment (Kitzunu) [inline src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp:80]: You can use Positions?

Comment (Kitzunu) [reply src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp:80]: Position const elementalPos[4]

Comment (elthehablo) [reply src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp:80]: In the DB you mean?

Comment (elthehablo) [reply src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp:80]: Summoncreature uses floats unfortunately

Comment (elthehablo) [reply src/server/scripts/EasternKingdoms/Karazhan/boss_shade_of_aran.cpp:80]: Or is it overloaded? I think it's not

**Review by Nyeriah — APPROVED** [review]

**Pull request merged**